File: dynamo_ts/src/basket/addToBasket.ts
import { DynamoDB } from 'aws-sdk';
import { HttpRequest, HttpResponse } from '../http'
import { Basket, BasketProduct, SimpleProduct, validateSimpleProduct } from './forms';

const ddb = new DynamoDB.DocumentClient({ region: "eu-central-1" })

export async function handler(event: HttpRequest): Promise<HttpResponse> {
  const simpleProduct: SimpleProduct = JSON.parse(event.body)
  const validationErr = validateSimpleProduct(simpleProduct)

  if (validationErr) {
    return {
      statusCode: 400,
      body: validationErr
    }
  }

  
  try {
    const updatedBasket = await addToBasket(event.pathParameters['userId'], simpleProduct)
    const params = {
      TableName: process.env.BASKET_TABLE!,
      Item: updatedBasket
    }
    const data = await ddb.put(params).promise()
    if (data.Attributes) {
      return {
        statusCode: 200,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Credentials': 'true',
        },
        body: JSON.stringify(data.Attributes)
      }
    }
    return {
      statusCode: 404,
      body: "Item not found"
    }
  } catch (err) {
    return {
      statusCode: err.statusCode || 500,
      body: "Internal Server Error Oops: " + err
    }
  }
}

async function addToBasket(userId: string, product: SimpleProduct): Promise<Basket> {
  const params = {
    TableName: process.env.BASKET_TABLE!,
    Key: {
      userId
    }
  }
  const data = await ddb.get(params).promise()
  const basket = data.Item ? data.Item as Basket : { userId, products: [] }
  const e = basket.products.find(element => element.category === product.category && element.productId === product.productId);
  if (e) {
    e.amount = e.amount + 1
  } else {
    basket.products.push({
      category: product.category,
      productId: product.productId,
      amount: 1
    })
  }
  return basket
}

File: dynamo_go/lambda/basket/addToBasket/addToBasket.go
package main

import (
	"encoding/json"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
	"os"
	"webshop/internal/apigw"
	"webshop/internal/forms"
)

func Handler(event *events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {
	ddb := dynamodb.New(session.New())

	userId := event.PathParameters["userId"]

	var simpleProduct forms.SimpleProduct
	if err := json.Unmarshal([]byte(event.Body), &simpleProduct); err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	basket, err := addToBasket(userId, simpleProduct, ddb)
	if err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	item, err := dynamodbattribute.MarshalMap(basket)
	if err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	_, err = ddb.PutItem(&dynamodb.PutItemInput{
		TableName: aws.String(os.Getenv("BASKET_TABLE")), //aws.String macht wert zu einem  pointer
		Item:      item,
	})

	if err != nil {
		return apigw.CreateResponse(500, err.Error()), err
	}

	return apigw.CreateResponse(200, "Added one of "+event.Body+" to basket"), err
}

func addToBasket(id string, product forms.SimpleProduct, ddb *dynamodb.DynamoDB) (*forms.Basket, error) {
	basket := new(forms.Basket)
	out, err := ddb.GetItem(&dynamodb.GetItemInput{
		TableName: aws.String(os.Getenv("BASKET_TABLE")), //aws.String macht wert zu einem pointer
		Key: map[string]*dynamodb.AttributeValue{
			"userId": {S: aws.String(id)},
		},
	})
	if err != nil {
		return basket, err
	}
	//if basket is present - unmarshal and search for product
	productAlreadyThere := false
	if len(out.Item) > 0 {
		if err := dynamodbattribute.UnmarshalMap(out.Item, basket); err != nil {
			return basket, err
		}
		for i, p := range basket.Products {
			if p.ProductId == product.ProductId && p.Category == product.Category {
				productAlreadyThere = true
				basket.Products[i].Amount += 1
			}
		}
	}
	if !productAlreadyThere {
		basketProduct := new(forms.BasketProduct)
		basketProduct.ProductId = product.ProductId
		basketProduct.Category = product.Category
		basketProduct.Amount = 1
		basket.UserId = id
		basket.Products = append(basket.Products, basketProduct)
	}
	return basket, nil
}

func main() {
	lambda.Start(Handler)
}
